master/internal/trial.go

```diff
@@ -547,7 +561,19 @@ func (t *trial) processCompletedWorkload(ctx *actor.Context, msg searcher.Comple
 		}
 	}
 
+	// If we completed a searcher operation, synchronize with the searcher to allow it to relay any
+	// new operations to the trial. Otherwise just continue the trial immediately.
+	if completedSearcherOp {
+		ctx.Tell(ctx.Self().Parent(), continueTrial{trialID: t.id})
+		return nil
+	}
+	return t.continueTrial(ctx)
+}
+
+func (t *trial) continueTrial(ctx *actor.Context) error {
```

Comment:
Two things.  First, `continueTrial` needs to be renamed, since that name belongs to a completely different concept in our system.

Second, By using a pair of `Tell`s instead of an `Ask`, we can now have a situation where a Trial is in the middle of processing a completed workload (after the first Tell) and the experiment has already given the trial more workloads to complete before the second Tell.  This would arise in ASHA, right?

At that point, you would enter this `continueTrial()` function after you may have already received other commands which affect the state.

The seemingly worst case of this that I can think of is if a taskTerminated message came which triggered a `resetTrial` before the experiment's `Tell(continueTrial)` made it back to the trial.  But trial.go is so complicated I can't really tell if that would automagically be handled correctly or not.  Without spending a great deal more time looking at it.

Have you already walked through that possibility?  Do you know that it is safe?

Reply:
> we can now have a situation where a Trial is in the middle of processing a completed workload (after the first Tell) and the experiment has already given the trial more workloads to complete before the second Tell. This would arise in ASHA, right?

This behavior is essentially the intent of this change. We want the experiment to finish processing any operations created by trial.go line 543/551 _before_ we get to the decision to close the trial out or not.

> The seemingly worst case of this that I can think of is if a taskTerminated message came which triggered a resetTrial before the experiment's Tell(continueTrial) made it back to the trial. But trial.go is so complicated I can't really tell if that would automagically be handled correctly or not. Without spending a great deal more time looking at it.

Because of how we changed the checkpointing logic, we'll never rollback past a searcher decision and end up in an invalid state.

Reply:
>  First, continueTrial needs to be renamed, since that name belongs to a completely different concept in our system.

damn..